tests/test_snql_api.py

```diff
@@ -1391,6 +1391,75 @@ def test_hexint_in_condition(self) -> None:
         data = json.loads(response.data)
         assert "9533608433997996441" in data["sql"], data["sql"]  # Hexint was applied
 
+    def test_allocator_clickhouse_bug(self) -> None:
+        response = self.post(
+            "/discover/snql",
+            data=json.dumps(
+                {
+                    "dataset": "events",
+                    "query": f"""
+                    MATCH (discover)
+                    SELECT
+                        divide(count(), divide(320519.0, 60)) AS `epm`,
+                        quantile(0.5)(duration) AS `p50`,
+                        countIf(notIn(transaction_status, tuple(2, 0, 1))) AS `failure_count`,
+                        transaction BY transaction
+                    WHERE
+                        type = 'transaction'
+                        AND release IN tuple('1123581321345589')
+                        AND environment IN array('prod', 'dev')
+                        AND timestamp >= toDateTime('{self.base_time.isoformat()}')
+                        AND timestamp < toDateTime('{self.next_time.isoformat()}')
+                        AND project_id IN tuple({self.project_id})
+                    HAVING
+                        countIf(notIn(transaction_status, tuple(2, 0, 1))) AS `failure_count` > 0.0
+                    ORDER BY
+                        countIf(notIn(transaction_status, tuple(2, 0, 1))) AS `failure_count` DESC
+                    LIMIT
+                        6 OFFSET 0""",
+                    "legacy": True,
+                    "app_id": "legacy",
+                    "tenant_ids": {
+                        "organization_id": self.org_id,
+                        "referrer": "join.tag.test",
+                    },
+                    "parent_api": "/api/0/issues|groups/{issue_id}/tags/",
+                }
+            ),
+        )
+
+        assert response.status_code == 200
+        data = json.loads(response.data)
+        assert "equals((release AS _snuba_release), '1123581321345589')" in data["sql"]
+        assert (
+            "has(['prod', 'dev'], (environment AS _snuba_environment))" in data["sql"]
+        )
+
+    def test_discover_semver_bug(self) -> None:
+        response = self.post(
+            "/discover/snql",
+            data=json.dumps(
+                {
+                    "consistent": False,
+                    "turbo": False,
+                    "query": "MATCH (discover) SELECT event_id AS `id` WHERE release IN array('test@1.2.3+123', 'test2@1.2.4+124') AND timestamp >= toDateTime('2024-04-18T22:28:40.928000') AND timestamp < toDateTime('2024-04-19T22:28:40.306000') AND project_id IN array(4553863597588481) LIMIT 50",
```

Comment:
@evanh this is the sentry test that's failing in a bizarre fashion:

```
'DB::Exception: First argument for function has must be an array.. 
```

sql:

```
SELECT (replaceAll(toString(event_id), '-', '') AS _snuba_event_id) FROM discover_local 
PREWHERE in((project_id AS _snuba_project_id), [4553863597588481]) 
WHERE equals(deleted, 0) AND has(['test@1.2.3+123', 'test2@1.2.4+124'], (release AS _snuba_release)) 
AND greaterOrEquals((timestamp AS _snuba_timestamp), toDateTime('2024-04-18T22:28:40', 'Universal')) 
AND less(_snuba_timestamp, toDateTime('2024-04-19T22:28:40', 'Universal')) LIMIT 50 OFFSET 0
```

Reply:
Can you replace all the hardcoded values here with test values (project_id, org_id, timestamp etc.)?

Reply:
these are the values that are in the failing sentry test

Reply:
Just tried it, it fails the same way

Reply:
This is very weird, because the SQL query works in production and when running it directly on Clickhouse.

Reply:
Had a (hacky) idea: I just removed the optimizer from the discover entity. We don't actually need it there, just discover_transactions and transactions.